src/actions/expressions.js

```diff
@@ -37,26 +38,30 @@ export function addExpression(input: string) {
       return dispatch(evaluateExpression(expression));
     }
 
-    dispatch({
-      type: "ADD_EXPRESSION",
-      input
-    });
+    dispatch({ type: "ADD_EXPRESSION", input, expressionError });
 
     const newExpression = getExpression(getState(), input);
-    dispatch(evaluateExpression(newExpression));
+    if (newExpression) {
+      return dispatch(evaluateExpression(newExpression));
+    }
   };
 }
 
+export function clearExpressionError() {
+  return { type: "CLEAR_EXPRESSION_ERROR" };
+}
+
 export function updateExpression(input: string, expression: Expression) {
-  return ({ dispatch, getState }: ThunkArgs) => {
+  return async ({ dispatch, getState }: ThunkArgs) => {
+    const expressionError = await parser.hasSyntaxError(input);
     if (!input || input == expression.input) {
       return;
     }
-
     dispatch({
       type: "UPDATE_EXPRESSION",
       expression,
-      input: input
+      input: expressionError ? expression.input : input,
+      expressionError
```

Comment:
maybe we want to have a separate action for `setExpressionError`

```js
if (error) {
  dispatch(setExpressionError(..))
else 
  dispatch(
    type: "UPDATE_EXPRESSION",
    ...
  )
```

the advantage is that that will simplify update expression. And we can do the same thing as w/ add expression where if we are updating the expression we can clear the error...

Reply:
I had the additional action before and the reason I have the line above isn't for setting the expression it's rather so we don't change the content of the expression to the new invalid value.